src/main/java/net/dv8tion/jda/api/requests/Route.java

```diff
@@ -329,100 +315,317 @@ public static Route custom(@Nonnull Method method, @Nonnull String route)
+     *
+     * @return The compiled route, ready to use for rate-limit handling
+     */
+    @Nonnull
+    public CompiledRoute compile(@Nonnull String... params)
     {
-        if (params.length != paramCount)
-        {
-            throw new IllegalArgumentException("Error Compiling Route: [" + route + "], incorrect amount of parameters provided." +
-                    "Expected: " + paramCount + ", Provided: " + params.length);
-        }
-
-        //Compile the route for interfacing with discord.
-        Set<String> major = new HashSet<>();
-        StringBuilder compiledRoute = new StringBuilder(route);
-        for (int i = 0; i < paramCount; i++)
+        Checks.noneNull(params, "Arguments");
+        Checks.check(
+            params.length == paramCount,
+            "Error Compiling Route: [%s], incorrect amount of parameters provided. Expected: %d, Provided: %d",
+            this, paramCount, params.length
+        );
+
+        StringJoiner major = new StringJoiner(":").setEmptyValue("n/a");
+        StringJoiner compiledRoute = new StringJoiner("/");
+
+        int paramIndex = 0;
+        for (String element : template)
         {
-            int paramStart = compiledRoute.indexOf("{");
-            int paramEnd = compiledRoute.indexOf("}");
-            String paramName = compiledRoute.substring(paramStart+1, paramEnd);
-            if (majorParameters.contains(paramName))
+            if (element.charAt(0) == '{')
             {
-                if (params[i].length() > 30) // probably a long interaction_token, hash it to keep logs clean (not useful anyway)
-                    major.add(paramName + "=" + Integer.toUnsignedString(params[i].hashCode()));
-                else
-                    major.add(paramName + "=" + params[i]);
+                String name = element.substring(1, element.length() - 1);
+                String value = params[paramIndex++];
+                if (MAJOR_PARAMETER_NAMES.contains(name))
+                {
+                    if (value.length() > 30) // probably a long interaction_token, hash it to keep logs clean (not useful anyway)
+                        major.add(name + "=" + Integer.toUnsignedString(value.hashCode()));
+                    else
+                        major.add(name + "=" + value);
+                }
+                compiledRoute.add(EncodingUtil.encodeUTF8(value));
+            }
+            else
+            {
+                compiledRoute.add(element);
             }
-
-            compiledRoute.replace(paramStart, paramEnd + 1, params[i]);
         }
 
-        return new CompiledRoute(this, compiledRoute.toString(), major.isEmpty() ? "n/a" : String.join(":", major));
+        return new CompiledRoute(this, compiledRoute.toString(), major.toString());
     }
 
     @Override
     public int hashCode()
     {
-        return (route + method.toString()).hashCode();
+        return 31 * Arrays.hashCode(template) + method.hashCode();
```

Comment:
Can we not use some combination of `Objects.hashCode` or `Objects.hash` here? direct usage of `31 *` feels weird.

Reply:
It does the exact same but sure